include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,245 @@
+#pragma once
+
+#include <mutex>
+#include <shared_mutex>
+
+#include <rclcpp>
+
+namespace rosparam_shortcuts {
+
+/*
+ * Thread-safe container for generic parameter values, providing callback functions.
+ * This class implements a tree structure of parameters which are declared recursively.
+ * All leafs are of type Parameter<rclcpp::Parameter> which are used for interacting with
+ * the rclcpp node parameter interface using the NodeConfig class further down.
+ */
+template <typename ParameterT>
+class Parameter
+{
+public:
+	Parameter(const std::string& name);
+	Parameter(const std::string& name, const ParameterT& default_value);
+
+	// Fields
+	const std::string& name() const noexcept;
+	std::string description() const noexcept;
+	std::string additional_constraints() const noexcept;
+	bool isNodeParameter() const noexcept;  // true if this parameter contains an rclcpp::Parameter
+
+	// Properties
+	Parameter<ParameterT>& describe(const std::string& description, const std::string& additional_constraints = "");
+	// Allow setting additional properties. i.e. read_only, dynamic_typing?
+	// void setProperties(...);
+
+	// Validity Status
+	const bool isValid() const noexcept;  // errors empty?
+	const bool hasErrors() const noexcept;
+	const bool hasWarnings() const noexcept;
+	// TODO: Insert additional thread-safe access to errors/warnings
+
+	// Read/write-locked data access using std::shared_mutex
+	// See: https://www.youtube.com/watch?v=ozOgzlxIsdg
+	ReadLockedPtr<ParameterT> operator->const;
+	WriteLockedPtr<ParameterT> operator->;
+	// optional: (useful for synchronized vector/map operations)
+	void applyRW(std::function<void(ParameterT& value)>);
+	void applyRO(std::function<void(const ParameterT& value)>);
+
+	// Copy-set/Copy-get value using read/write lock internally
+	ParameterT& operator=(const ParameterT& value);
+	ParameterT operator() const;  // or:  'operator ParameterT() const;'
+
+	// Validity Callbacks (could also use rcl_interfaces::msg::SetParametersResult like in node_parameters.h).
+	// Callbacks are stored in ordered hash map, could also possibly return hash for deleting later on.
+	// Alternatively, return Parameter<ParameterT>& for method chaining
+	using ParameterValidityCallback = std::function<bool(const ParameterT& value, std::string& message)>;
+	Parameter<ParameterT>& warnIf(ParameterValidityCallback warn_if_callback);
+	Parameter<ParameterT>& errorIf(ParameterValidityCallback error_if_callback);  // or rejectIf()/failIf()..
```

Comment:
I love the idea to support these predicates in the API.

I would consider to add the textual description together with each predicate here, instead of describing them independently in `describe`. See my example below.

Reply:
So.. `describe()` doesn't really describe the predicates, it describes the parameter itself. I agree that we should extend the API to allow setting the predicate descriptions like you suggested below.

Reply:
Well, the `additional_constraints` field is currently set through `describe()` and should pretty much describe the predicates on these parameters. That's what I meant. :-)